File: src/main/java/knapsack/java/Knapsack.java
package knapsack.java;

import java.util.LinkedList;
import java.util.List;
import java.util.stream.Collectors;

public class Knapsack {

    private int capacity = 50;
    private List<Item> items = new LinkedList<>();

    public Knapsack(int capacity) {
        this.capacity = capacity;
    }

    public int calculateValue() {
        return this.items.stream().mapToInt(Item::getValue).sum();
    }

    public int calculateWeight() {
        return this.items.stream().mapToInt(Item::getWeight).sum();
    }

    public double calculateValueDividedByWeight() {
        return calculateValue() / calculateWeight();
    }

    public boolean willExceedSize(Item item) {
        return this.calculateWeight() + item.getWeight() > capacity;
    }

    public void put(Item item) {
        items.add(item);
    }

    public boolean isFull() {
        return this.calculateWeight() == capacity;
    }

    public Knapsack copyWithItem(Item item) {
        Knapsack knapsack = new Knapsack(this.capacity);
        List<Item> newItems = this.items.stream()
                .map(it -> new Item(it.getValue(), it.getWeight()))
                .collect(Collectors.toList());
        newItems.forEach(knapsack::put);
        knapsack.put(item);
        return knapsack;
    }

    public void printSummary() {
        System.out.println("Knapsack with capacity " + capacity + " has items: ");
        this.items.forEach(System.out::println);
    }

    public boolean canBeAddedAnyOf(List<Item> items) {
        return items.stream().anyMatch(it -> !willExceedSize(it));
    }

    public List<Item> getItems() {
        return items;
    }

    @Override
    public String toString() {
        return "Knapsack{" +
                "capacity=" + capacity +
                ", items=" + items +
                '}';
    }
}

File: src/main/java/knapsack/Knapsack.kt
package knapsack

import java.util.*

data class Knapsack(
        private val capacity: Int = 50
) {
    val items: MutableList<Item> = LinkedList()

    fun calculateValue() = items.sumBy { it.value }

    fun calculateValueDividedByWeight() = calculateValue() / calculateWeight()

    fun calculateWeight() = items.sumBy { it.weight }

    fun willExceedSize(item: Item) = this.calculateWeight() + item.weight > capacity

    fun put(item: Item) = items.add(item)

    fun isFull() = this.calculateWeight() == capacity

    fun copyWithItem(item: Item): Knapsack {
        val knapsack = Knapsack(capacity)
        this.items.forEach { knapsack.put(it) }
        knapsack.put(item)
        return knapsack
    }

    fun printSummary() {
        println("Knapsack with capacity = $capacity has items:")
        val sorted = items.sortedByDescending { it.value }
        sorted.forEach {
            println("Item: v=${it.value} w=${it.weight}")
        }
    }

    fun canBeAddedAnyOf(items: MutableList<Item>): Boolean {
        return items.any { !willExceedSize(it) }
    }
}

data class Item(
        val value: Int,
        val weight: Int
)

